### DISCUSSION TITLE
Add github release backend

### DISCUSSION THREAD

**Discussion by iilyak** [jdx/mise#1686] [Ideas]
Quite often modern software is distributed as github releases. Lot's of asdf plugins interact with github to get list of available versions and download/extract artifacts. The differences between plugins really is about:

- location of the repo
- naming conventions of the artifact
- naming convention of the platform
- naming conventions of the version
- differences in archive type
- difference in permissions of the files after extraction

There are a bunch of generic "installers" of various quality which provide ability to download arbitrary tools. 

It would be grade if instead of repeating the same scripts in asdf plugin there would be a way to download directly from github. This would need to support verification of a checksum.

**Comment by iilyak** [comment]
Few projects as a source of inspiration for UX

- https://github.com/giomf/grpm
- https://github.com/chshersh/tool-sync
- https://github.com/egoist/bina
- https://github.com/xermicus/blindspot
- https://github.com/cjrh/lifter
- https://github.com/marcosnils/bin
- https://github.com/izirku/gitrel
- https://github.com/zyedidia/eget
- https://github.com/jpillora/installer
- https://github.com/innobead/huber
- https://github.com/egoist/bget
- https://github.com/jooola/gh-release-install
- https://github.com/sheepla/ghin
- https://github.com/warrensbox/hubapp
- https://github.com/TrendingTechnology/instl
- https://github.com/installer/instl
- https://github.com/Rishang/install-release
- https://github.com/zloeber/ghr-installer
- https://github.com/Songmu/ghg

Notable mentions which are not really installers

- https://aquaproj.github.io/docs/develop-registry/add-registry

**Comment by iilyak** [comment]
Huber developers were able to generate the registry. If that would be possible then asdf plugins wouldn't be needed at all (for software which is distributed as single binary or simple archive).

**Comment by iilyak** [comment]
- https://github.com/DISTREAT/dyst
- https://github.com/marwanhawari/stew

**Comment by jdx** [comment]
eget is the one I looked at a bit in https://github.com/jdx/mise/discussions/1250.

There are so many tools in this space I don't like the idea of mise taking on the actual work of integrating with github releases directly. I think the right way to do this is via a backend for something like eget.

**Reply by iilyak** [reply]
Fair enough

**Reply by iilyak** [reply]
One more https://github.com/aquaproj/aqua (which I'll probably end up using).

**Reply by iilyak** [reply]
Although my project setup is getting hairy:

- aquaproj for github tools
- https://github.com/go-task/task/ - as a task runner
- mise as a replacement for direnv+asdf combination I previously used.

**Reply by jdx** [reply]
I'd be curious where using mise for tasks isn't suitable for you

**Reply by iilyak** [reply]
> I'd be curious where using mise for tasks isn't suitable for you

The features I need (not all supported by go-task)

1. chain tasks in a way so I can register output of one task under some name and use it as argument in another task.
2. rendering templates using data from taskfile (or loaded data-file json/yaml) to files
3. prevent a task from running if a given set of files haven't changed since last run
4. remote taskfiles
5. management of secrets (removing them from the stdout/stderr output) when I pass them to shell command

**Reply by jdx** [reply]
I know [moon](https://github.com/moonrepo/moon) supports some of these, I haven't used it but might be worth a look. I think a lot of these are probably out of scope for mise since I see it really more as a "task runner" and not a "build system", though the 2 are very similar. I could see people using something like mise on top of a more comprehensive build system too.

**Reply by jdx** [reply]
out of what you listed, 5 is almost certainly something mise _will_ support at some point though I'm really unsure exactly how secrets will work. It already supports 3. 2 I think could happen but I'm not sure. I don't think 1 or 4 makes sense for mise though.

I have thought that a "remote mise" might be an interesting idea, I've been thinking more as an ansible/chef kind of thing. This isn't something I think I would try to tackle on any near timeframe though.

**Reply by airtonix** [reply]
> chain tasks in a way so I can register output of one task under some name and use it as argument in another task.

I raised https://github.com/jdx/mise/issues/2564 to address this point

**Reply by airtonix** [reply]
> prevent a task from running if a given set of files haven't changed since last run

Could be handled by a separate tool that is given globs or array of files and fingerprints/inspects them into some `/tmp/blah` and exit 1 or 0 based on outcome.

**Comment by olfway** [comment]
Aqua projects publish their registry on github, probably it also could be used by mise
https://github.com/aquaproj/aqua-registry/blob/main/registry.yaml
